framework/src/arastorage/index_bplustree.c

```diff
@@ -2002,3 +2023,560 @@ static int db_flush(tree_t *tree, relation_t *rel)
+		rsb_id = pn->id[index + 1];
+		rsbn = tree_read(tree, rsb_id);
+		if (rsbn && (rsbn->val[BRANCH_FACTOR - 1] > BRANCH_FACTOR / 2)) {
+			//move key, and child id, from sibling and parent
+			key_num = n->val[BRANCH_FACTOR - 1];
+			n->val[key_num] = pn->val[index];
+			n->id[key_num+1] = rsbn->id[0];
+			pn->val[index] = rsbn->val[0];
+			n->val[BRANCH_FACTOR - 1] = key_num + 1;
+
+			//update sibling keys
+			sb_key_num = rsbn->val[BRANCH_FACTOR - 1];
+			for (i = 0; i < sb_key_num; i++) {
+				rsbn->val[i] = rsbn->val[i + 1];
+				rsbn->id[i] = rsbn->id[i + 1];
+			}
+			rsbn->val[BRANCH_FACTOR - 1] = sb_key_num - 1;
+
+			modify_cache(tree, node_id, NODE, UNLOCK);
+			modify_cache(tree, pnode_id, NODE, UNLOCK);
+			modify_cache(tree, rsb_id, NODE, UNLOCK);
+			if (bmerge_left) {
+				modify_cache(tree, lsb_id, NODE, UNLOCK);
+			}
+
+			return 0;
+		}
+
+		bmerge_right = (rsbn != NULL);
+	}
+
+	//need to merge node
+	if (bmerge_left || bmerge_right) {
+		//merge to left node
+		key_num = n->val[BRANCH_FACTOR - 1];
+		sb_key_num = bmerge_left ? lsbn->val[BRANCH_FACTOR - 1] : rsbn->val[BRANCH_FACTOR - 1];
+		if (bmerge_left) {
+			lsbn->val[sb_key_num++] = pn->val[index - 1];
+			for (i = 0; i < key_num; i++) {
+				lsbn->val[sb_key_num + i] = n->val[i];
+				lsbn->id[sb_key_num + i] = n->id[i];
+			}
+			lsbn->id[sb_key_num + key_num] = n->id[key_num];
+			lsbn->val[BRANCH_FACTOR - 1] = sb_key_num + key_num;
+		} else {
+			key_num++;
+			for (i = sb_key_num; i >= 0 ; i--) {
+				rsbn->val[key_num + i] = rsbn->val[i];
+				rsbn->id[key_num + i] = rsbn->id[i];
+			}
+			for (i = 0; i < key_num; i++) {
+				rsbn->val[i] = n->val[i];
+				rsbn->id[i] = n->id[i];
+			}
+			rsbn->val[key_num-1] = pn->val[index];
+			rsbn->val[BRANCH_FACTOR - 1] = sb_key_num + key_num;
+		}
+
+		//release current node
+		modify_cache(tree, node_id, NODE, INVALIDATE);
+		tree->off_nodes--;
+
+		//update parent node
+		if (pn->val[BRANCH_FACTOR - 1] == 1) {  //parent node MUST be root
+			tree->root = bmerge_left ? lsb_id : rsb_id;
+			DB_LOG_D("[tree->root]tree root change to %d\n", tree->root);
+			tree->levels--;
+			tree->off_nodes--;
+			modify_cache(tree, pnode_id, NODE, INVALIDATE);
+		} else {
+			if (bmerge_left) {
+				pn->val[index - 1] = pn->val[index];
+			}
+			for (i = index; i < pn->val[BRANCH_FACTOR - 1]; i++) {
+				pn->val[i] = pn->val[i + 1];
+				pn->id[i] = pn->id[i + 1];
+			}
+			pn->val[BRANCH_FACTOR - 1] = pn->val[BRANCH_FACTOR - 1] - 1;
+
+			if ((pnode_id != tree->root) && pn->val[BRANCH_FACTOR - 1] < BRANCH_FACTOR / 2) {
+				tree_rebuild_node(tree, path, level-1);
+			}
+		}
+	}
+
+	if (bmerge_left) {
+		modify_cache(tree, lsb_id, NODE, UNLOCK);
+	}
+	if (bmerge_right) {
+		modify_cache(tree, rsb_id, NODE, UNLOCK);
+	}
+
+	modify_cache(tree, node_id, NODE, UNLOCK);
+	modify_cache(tree, pnode_id, NODE, UNLOCK);
+
+	return 0;
+}
+
+static int bucket_merge_sibling(tree_t *tree, pair_t *path, bucket_t *bucket, int rm_val)
+{
```

Comment:
No input pointer check. if "tree" is null, it will crash.

Reply:
it's internal function which could guarantee this